### ISSUE TITLE
Zero HLT bits count in EventMessage

### ISSUE THREAD

**Issue by pmandrik** [cms-sw/cmssw#37029]
As described at [0] the DQM clients with trigger preselection crashed during MWGR with following error: 

```
An exception of category 'StdException' occurred while
   [0] Calling InputSource::getNextItemType
Exception Message:
A std::exception was thrown.
vector::_M_range_check: __n (which is 1) >= this->size() (which is 0)
```

The crashed line is in the /FWCore/Framework/src/EventSelector.cc [1] trigger selector. The reason for this is mismatch between number of HLT bits taken from /IOPool/Streamer/src/EventMessage.cc and trigger names available in the /IOPool/Streamer/src/InitMessage.cc. To be more precise, in the MWGR stream files number of HLT bits is zero ( hlt_bits_count_ at [2] line ).

The instruction to reproduce:
```
scram project CMSSW_12_2_1
cd CMSSW_12_2_1/src; cmsenv
cp -r /afs/cern.ch/user/p/pmandrik/public/temp/data_hlt_bits data
cmsRun DQM/Integration/python/clients/dt_dqm_sourceclient-live_cfg.py runInputDir=./data runNumber=347814 runkey=cosmic_run scanOnce=True
```

[0] http://cmsonline.cern.ch/cms-elog/1131038
[1] https://github.com/cms-sw/cmssw/blob/7d5358ff0e31e78fbcb737c05d218a55f5d82104/FWCore/Framework/src/EventSelector.cc#L373
[2] https://github.com/cms-sw/cmssw/blob/2ee2dccbf040d3c0a5ad49527b5615f710a07e34/IOPool/Streamer/src/EventMessage.cc#L48

**Comment by makortel** [comment]
Is there a way to reproduce the file creation? It seems to me the problem is in the file itself (but I could be wrong here), and the change @smorovic mentioned could certainly play a role here.

**Comment by smorovic** [comment]
I will try to reproduce it with generated data (based on unit test in EventFilter/Utilities/test).

**Comment by smorovic** [comment]
I could reproduce it. You can copy this directory: `/afs/cern.ch/user/s/smorovic/public/test_outputmodule`
and run with 12_2_1 environment:
```
./do_run.sh
```
when GlobalEvFOutputModule is replaced by the EvFOutputModule in `hlt_test.py`, there is no longer an exception, so it's confirmed that it is related to the new module.

**Comment by smorovic** [comment]
Issue is that hltsize_ is 2 here:
https://github.com/cms-sw/cmssw/blob/master/IOPool/Streamer/src/StreamerOutputModuleCommon.cc#L124

but is 0 here:
https://github.com/cms-sw/cmssw/blob/master/IOPool/Streamer/src/StreamerOutputModuleCommon.cc#L166

First happens during beginRun, the latter is on event processing.
I can see that constructor of 'StreamerOutputModuleCommon`, which initializes hltsize_ to 0, gets called 3 times, but serializeRegistry only once.

With EvFOutputModule everything is called only once.

**Comment by makortel** [comment]
I think I found the problem. The number of HLT bits is written into the InitMessage and EventMessage by `edm::StreamerOutputModuleCommon`. The number of the HLT bits is read in `edm::StreamerOutputModuleCommon::serializeRegistry()`
https://github.com/cms-sw/cmssw/blob/69c0b8342053669d51c832c0f5053a8e796a5a50/IOPool/Streamer/src/StreamerOutputModuleCommon.cc#L123-L124
and then used in `StreamerOutputModuleCommon::serializeEvent()`
https://github.com/cms-sw/cmssw/blob/69c0b8342053669d51c832c0f5053a8e796a5a50/IOPool/Streamer/src/StreamerOutputModuleCommon.cc#L235-L246

The `GlobalEvFOutputModule` constructs the `edm::StreamerOutputModuleCommon` in `beginStream()` for `StreamCache`
https://github.com/cms-sw/cmssw/blob/69c0b8342053669d51c832c0f5053a8e796a5a50/EventFilter/Utilities/plugins/GlobalEvFOutputModule.cc#L292-L294
and uses that to serialize the event data. But the InitMessage is serialized in `globalBeginRun()` with a **local** `edm::StreamerOutputModuleCommon` object
https://github.com/cms-sw/cmssw/blob/69c0b8342053669d51c832c0f5053a8e796a5a50/EventFilter/Utilities/plugins/GlobalEvFOutputModule.cc#L296-L319
that has its `hltsize_ == 0`.

**Comment by smorovic** [comment]
There is also `outputModuleId_`, which depends `description().moduleLabel(),` from the module. It could be passed as a constructor parameter.

**Comment by Dr15Jones** [comment]
> I wonder if the edm::StreamerOutputModuleCommon::hltsize_ could be just set properly in its constructor. It needs TriggerNameService for that, but that information should be available at module construction time.

I'm not sure when TriggerNameService is actually fully populated. Modules are being constructed while each Path is being made.

**Comment by makortel** [comment]
> > I wonder if the edm::StreamerOutputModuleCommon::hltsize_ could be just set properly in its constructor. It needs TriggerNameService for that, but that information should be available at module construction time.
> 
> I'm not sure when TriggerNameService is actually fully populated. Modules are being constructed while each Path is being made.

I was concerned of that too, but the Path names appear to be set in the constructor of `TriggerNameService` from configuration
https://github.com/cms-sw/cmssw/blob/0a95f98a9f01ae9b9bd5db784b1dd047a42c1704/FWCore/Framework/src/getAllTriggerNames.cc#L22-L24
https://github.com/cms-sw/cmssw/blob/0a95f98a9f01ae9b9bd5db784b1dd047a42c1704/FWCore/Framework/interface/TriggerNamesService.h#L55
https://github.com/cms-sw/cmssw/blob/0a95f98a9f01ae9b9bd5db784b1dd047a42c1704/FWCore/Framework/src/TriggerNamesService.cc#L18-L21

**Comment by smorovic** [comment]
HI @ahmad3213 ,
I am preparing a fix.
The unit test is not yet ready, but the code is; I will push it now so that we can get it released in time for craft.

**Comment by makortel** [comment]
@smorovic Thanks for the PRs. Did I understand your comment https://github.com/cms-sw/cmssw/issues/37029#issuecomment-1049665178 correctly that you're also preparing a unit test for this case (for which a PR would be made later)?

**Comment by smorovic** [comment]
@makortel, yes, I will expand existing unit test, which tests the data-flow in HLT, to also run DQM streamer source on the data in the chain. This will come in another PR.

**Comment by jfernan2** [comment]
+1
For the records

**Comment by missirol** [comment]
+hlt

**Comment by makortel** [comment]
+core
